elastic/kibana
# [Security Solution] [Platform] Adds support for data views and runtime field mappings in rule creation, exceptions, and during execution

## Summary

Ref: https://github.com/elastic/kibana/issues/103587 and https://github.com/elastic/kibana/issues/97778

#### Overview

Users will now have the option of selecting a data view or an index pattern when creating rules. Data views will be available in the EQL, Query, Threat Indicator** , and Threshold rule types. 

If a rule is configured to query a data view, any runtime field mappings defined on that data view will also be available as fields to execute during the query and for any applicable override fields such as the severity override, risk score override, rule name override, and timestamp override.

** for querying source documents only. There is not support for threat indicator data views, only index patterns for now.

## Affected flows
### Rule Creation
- User can select to use index pattern, or data view
- Only selection of a single data view, not multiple, is supported
- Runtime fields associated with data view are viewable in preview
- Options to overwrite fields will also display runtime fields if relevant (timestamp override, risk score...)
<details>
<summary>Rule creation</summary>
<img width="1001" alt="eql_rule_with_data_view_and_runtime_field" src="https://user-images.githubusercontent.com/915763/169394807-c73fdbc0-b2c5-44bf-86fd-dca7506852da.png">
</details>

<details>
<summary>Rule Preview</summary>
<img width="1026" alt="eql_rule_preview" src="https://user-images.githubusercontent.com/915763/169394905-dd628e13-e19d-4733-b183-be07d0b91518.png">
</details>

### Rule Edit
- User can select to use index pattern, or data view
- Only selection of a single data view, not multiple, is supported
- Runtime fields associated with data view are viewable in preview
- 👀 If the selected data view no longer exists (like it was deleted), user is prompted with a warning, letting them know
  - To test create a rule with a data view, go to Stack Management --> Data Views --> delete the data view you selected, go back to the rule and click to edit
<details>
<summary>Rule Edit when error present</summary>
<img width="1050" alt="Screen Shot 2022-05-19 at 1 44 23 PM" src="https://user-images.githubusercontent.com/10927944/169401660-3ae64e9e-9764-4664-876b-20da658d50c9.png">
</details>

### Rule Details
- User can see what data view they've selected
- 👀 If the selected data view no longer exists (like it was deleted), user is prompted with an error, letting them know that rule execution failed
  - To test create a rule with a data view, go to Stack Management --> Data Views --> delete the data view you selected, go back to the rule
<details>
<summary>Rule details when error present</summary>
<img width="2292" alt="Screen Shot 2022-05-19 at 1 47 14 PM" src="https://user-images.githubusercontent.com/10927944/169401880-5ab4bf55-2bc6-4ea8-9026-8de4dd8eba7b.png">
</details>

### Rules Management
- User can select to bulk edit rule index patterns
    - If they select to apply changes to rules with data views, those rules will be updated as well
    - If they do not select to apply changes to rules with data views, those rules are not updated, but the rest are
- To test:
    - Create a rule with a data view and a rule with index patterns. Go to rule management and select both rules. Select 'Bulk actions' --> Add index patterns --> go through flow to add index pattern (don't select any boxes). The rule using index patterns should be updated, the rule using data views should not be.
    - Create a rule with a data view and a rule with index patterns. Go to rule management and select both rules. Select 'Bulk actions' --> Add index patterns --> go through flow to add index pattern --> check the box `Apply changes to rules configured with data views`. The rule using index patterns should be updated, the rule using data views should not be.
<details>
<summary>Bulk action flyout</summary>
<img width="675" alt="Screen Shot 2022-05-19 at 1 54 00 PM" src="https://user-images.githubusercontent.com/10927944/169403077-71a17570-40f4-4f94-8791-c24a4237036c.png">
</details>
<details>
<summary>Bulk action flyout success toast</summary>
<img width="402" alt="Screen Shot 2022-05-19 at 1 28 22 PM" src="https://user-images.githubusercontent.com/10927944/169403161-c1edd45b-c8b6-4ef1-9a29-6906474eb78e.png">
</details>

### Add rule exception modal
- If rule list is assigned to is using runtime fields, runtime fields show as a field option to add an exception for

#### Example
A user has a runtime field mapping they developed on the `logs-*` data view in **Discover**

<details><summary>Add runtime field in Discover</summary>

![data_view_runtime](https://user-images.githubusercontent.com/915763/169387839-fbb14ef4-9d2e-47c1-aa05-75fcab3fa751.gif)

</details>


And the user would now like to use that runtime field mapping as part of a query for an EQL rule (or any of the rules listed above)

<details><summary>Rule creation</summary>

<img width="1001" alt="eql_rule_with_data_view_and_runtime_field" src="https://user-images.githubusercontent.com/915763/169394807-c73fdbc0-b2c5-44bf-86fd-dca7506852da.png">

</details>

<details><summary>Rule Preview</summary>

<img width="1026" alt="eql_rule_preview" src="https://user-images.githubusercontent.com/915763/169394905-dd628e13-e19d-4733-b183-be07d0b91518.png">

</details>

<details><summary>Runtime field is populated in the preview alert</summary>

<img width="1108" alt="eql_rule_alert_preview" src="https://user-images.githubusercontent.com/915763/169394980-24cb18ec-a70e-43cd-89dc-ca25e822eef3.png">

</details>

### Checklist

Delete any items that are not applicable to this PR.




### For maintainers
